### PR TITLE
feat(analytics): introduce analytics table

### PR DESCRIPTION
**Pull request by winged**
The analytics table provides insight into Caluma's data by collecting
information from forms, cases, and so on.

### CODE REVIEW

**Comment by derrabauke** [comment]
Found another issue. When queryieng the backend for records via:
```graphql
query getTableData {
  analyticsTable (slug: "SomeExistingSlug") {
    resultData {
      ...ResultData
    }
  }
}
``` 
I get an error message in the resulting data:
```json
{
  "errors": [
    {
      "message": "int() argument must be a string, a bytes-like object or a number, not 'NoneType'",
      "locations": [
        {
          "line": 130,
          "column": 3
        }
      ],
      "path": [
        "analyticsTable",
        "resultData",
        "records"
      ]
    }
  ],
  "data": {
    "analyticsTable": {
      "resultData": {
        "records": null
      }
    }
  }
}
```
Might be some incomplete implementation, which is normal for WIP, but wanted to state this somehow.

**Comment by winged** [comment]
@fugal-dy @open-dynaMIX @czosel ready for review

File: caluma/caluma_analytics/simple_table.py [hunk 1]
```diff
@@ -0,0 +1,770 @@
...
+            return FormDocumentField(
+                parent=self,
+                identifier=question.slug,
+                form_slug=question.sub_form_id,
+                subform_level=self.subform_level + 1,
+                visibility_source=self.visibility_source,
+            )
+
+
+class FormAnswerField(AttributeField):
```
Comment (open-dynaMIX) [inline caluma/caluma_analytics/simple_table.py:440]: Consider renaming to `DocumentAnswerField`.

File: caluma/caluma_analytics/simple_table.py [hunk 2]
```diff
@@ -0,0 +1,770 @@
...
+                date_field=self.attr_name,
+                visibility_source=self.visibility_source,
+            )
+            return {
+                "year": extractor_field(identifier="year"),
+                "month": extractor_field(identifier="month"),
+                "weekday": extractor_field(identifier="weekday"),
+                "quarter": extractor_field(identifier="quarter"),
+            }
+        return {}
```
Comment (open-dynaMIX) [inline caluma/caluma_analytics/simple_table.py:483]: Duplicated code. See `AttributeField`.

File: caluma/caluma_analytics/sql.py [hunk 3]
```diff
@@ -0,0 +1,428 @@
...
+        Note: No filter (filters=None, or filters=[]) means no restriction.
+        As it doesn't really make sense to remove all values (and thus all rows)
+        from the resulting data, we treat both as "don't filter" in this case
+        """
+        if not filter_values:
+            return
+
+        filters_sql = ", ".join(
+            [
+                # hacky SQL_INJECTION_PRONE FIRST ATTEMPT
```
Comment (open-dynaMIX) [inline caluma/caluma_analytics/sql.py:106]: Leftover comment.

File: caluma/caluma_analytics/sql.py [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+
+        The tuple consists of:
+        * subquery/table
+        * alias
+        """
+        if isinstance(from_spec, str):
+            # by defintion, only "our own" table can be passed as string.
+            # subqueries MUST be query objects
+            return (from_spec, {}, self.query.outer_alias())
+        else:
```
Comment (open-dynaMIX) [inline caluma/caluma_analytics/sql.py:400]: `else` not needed.

File: caluma/caluma_analytics/tests/test_form_field_extraction.py [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+            "case_id": case2.pk,
+            "blablub": answers_dict2["top_question"].value,
+        },
+        {
+            "case_id": case3.pk,
+            "blablub": answers_dict3["top_question"].value,
+        },
+    ]
+    expected_data = [analyzed_output[idx] for idx in expect_output]
+    print(table.last_query)
```
Comment (open-dynaMIX) [inline caluma/caluma_analytics/tests/test_form_field_extraction.py:264]: Leftover print statement

File: caluma/caluma_analytics/tests/test_simple_table.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+from caluma.caluma_analytics.simple_table import SimpleTable
+
+
+@pytest.fixture
+def example_analytics(analytics_table):
+    analytics_table.fields.create(data_source="created_at", alias="created_at")
+    analytics_table.fields.create(data_source="created_at.quarter", alias="quarter")
+    analytics_table.fields.create(data_source="meta.foo", alias="foo")
+    analytics_table.fields.create(data_source="status", alias="statuuuuuus")
+    # TODO: re-add document fields once tings are stabilized
```
Comment (open-dynaMIX) [inline caluma/caluma_analytics/tests/test_simple_table.py:14]: Unfinished TODO.

File: caluma/caluma_analytics/tests/test_table.py [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+            "id": to_global_id(type(analytics_field).__name__, analytics_field.pk)
+        }
+    }
+
+    result = schema_executor(query, variable_values=input)
+    assert not result.errors
+
+    # Try to delete a now-nonexistent field
+    result = schema_executor(query, variable_values=input)
+    assert result.errors
```
Comment (open-dynaMIX) [inline caluma/caluma_analytics/tests/test_table.py:245]: Please assert that the right error happened.

File: caluma/caluma_analytics/tests/test_table.py [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+    """
+
+    input = {"input": {"slug": analytics_table.pk}}
+
+    result = schema_executor(query, variable_values=input)
+    assert not result.errors
+
+    # Try to delete a now-nonexistent table
+    result = schema_executor(query, variable_values=input)
+    assert result.errors
```
Comment (open-dynaMIX) [inline caluma/caluma_analytics/tests/test_table.py:266]: Please assert that the right error happened.

**Review by open-dynaMIX — CHANGES_REQUESTED** [review]
Awesome work!!! :rocket:

**Comment by winged** [comment]
Weird - Don't know why all those snapshots disappeared. Investigating...

**Comment by winged** [comment]
Merged as per today's discussion :-)

**Pull request merged**